File: js/lib/GmodNode.ts
import { Gmod } from "./Gmod";
import { Location, Locations } from "./Location";
import { GmodNodeDto } from "./types/GmodDto";
import { GmodNodeMetadata } from "./types/GmodNode";
import { VIS } from "./VIS";
import { VisVersion } from "./VisVersion";

export class GmodNode {
    private _id: string;
    private _visVersion: VisVersion;
    public readonly code: string;
    public readonly metadata: GmodNodeMetadata;
    private _location?: Location;
    private _children: GmodNode[];
    private _parents: GmodNode[];

    private constructor(previous: GmodNode);
    private constructor(
        id: string,
        visVersion: VisVersion,
        code: string,
        metadata: GmodNodeMetadata,
        location?: Location,
        parents?: GmodNode[],
        children?: GmodNode[]
    );
    private constructor(
        arg1: string | GmodNode,
        visVersion?: VisVersion,
        code?: string,
        metadata?: GmodNodeMetadata,
        location?: Location,
        parents?: GmodNode[],
        children?: GmodNode[]
    ) {
        if (arg1 instanceof GmodNode) {
            const previous = arg1;
            this._id = previous._id;
            this._visVersion = previous._visVersion;
            this.code = previous.code;
            this.metadata = previous.metadata;
            this._location = previous._location;
            this._children = previous._children;
            this._parents = previous._parents;
        } else {
            this._id = arg1;
            this._visVersion = visVersion!;
            this.code = code!;
            this.metadata = metadata!;
            this._location = location;
            this._children = children ?? [];
            this._parents = parents ?? [];
        }
    }

    public get visVersion() {
        return this._visVersion;
    }

    public get location() {
        return this._location;
    }

    public static createFromDto(visVersion: VisVersion, dto: GmodNodeDto) {
        const {
            code,
            category,
            type,
            name,
            commonName,
            commonDefinition,
            definition,
            installSubstructure,
            normalAssignmentNames,
            id = code,
        } = dto;

        return new GmodNode(
            id,
            visVersion,
            code,
            {
                category,
                type,
                name,
                commonName: commonName ?? undefined,
                definition: definition ?? undefined,
                commonDefinition: commonDefinition ?? undefined,
                installSubstructure: installSubstructure ?? undefined,
                normalAssignmentNames: new Map<string, string>(
                    Object.entries(normalAssignmentNames ?? {})
                ),
            },
            undefined
        );
    }

    public static create(
        id: string,
        visVersion: VisVersion,
        code: string,
        metadata: GmodNodeMetadata,
        location?: Location,
        parents?: GmodNode[],
        children?: GmodNode[]
    ) {
        return new GmodNode(
            id,
            visVersion,
            code,
            metadata,
            location,
            parents,
            children
        );
    }

    public withEmptyRelations(): GmodNode {
        return new GmodNode(
            this._id,
            this._visVersion,
            this.code,
            this.metadata,
            this._location,
            [],
            []
        );
    }

    public withoutLocation(): GmodNode {
        return this.with((s) => (s._location = undefined));
    }

    public withLocation(
        location: string | undefined,
        locations: Locations
    ): GmodNode;
    public withLocation(location: Location): GmodNode;
    public withLocation(location: unknown, locations?: unknown): GmodNode {
        let result: GmodNode | undefined;
        if (location instanceof Location) {
            result = this.tryWithLocation(location);
        } else if (
            typeof location === "string" &&
            locations instanceof Locations
        ) {
            result = this.tryWithLocation(location, locations);
        }

        if (!result) throw new Error("Node not individualizable");
        return result;
    }

    public tryWithLocation(location: Location): GmodNode | undefined;
    public tryWithLocation(
        location: string | undefined,
        locations: Locations
    ): GmodNode | undefined;
    public tryWithLocation(
        location: unknown,
        locations?: unknown
    ): GmodNode | undefined {
        // if (!isIndividualizable(this, false, false)) return undefined; // TODO what to do about this case
        if (location instanceof Location) {
            return this.with((s) => (s._location = location));
        } else if (
            typeof location === "string" &&
            locations instanceof Locations
        ) {
            const parsedLocation = locations.tryParse(location);
            if (!parsedLocation) return undefined;
            return this.with((s) => (s._location = parsedLocation));
        }
    }

    public async withLocationAsync(location: string): Promise<GmodNode> {
        const locations = await VIS.instance.getLocations(this._visVersion);
        const parsedLocation = locations.parse(location);
        return this.withLocation(parsedLocation);
    }
    public async tryWithLocationAsync(
        location?: string
    ): Promise<GmodNode | undefined> {
        const locations = await VIS.instance.getLocations(this._visVersion);
        return this.tryWithLocation(location, locations);
    }

    public get id() {
        return this._id;
    }

    public get children() {
        return this._children;
    }

    public get parents() {
        return this._parents;
    }

    public get isMappable() {
        if (!!this.productType) return false;
        if (!!this.productSelection) return false;
        if (this.isProductSelection) return false;
        if (this.isAsset) return false;

        const lastChar = this.code.charAt(this.code.length - 1);
        return lastChar !== "a" && lastChar !== "s";
    }

    public get isProductSelection() {
        return Gmod.isProductSelection(this.metadata);
    }

    public get isProductType() {
        return Gmod.isProductType(this.metadata);
    }

    public get isAsset() {
        return Gmod.isAsset(this.metadata);
    }

    public get productType(): GmodNode | undefined {
        if (this._children.length !== 1) return;
        if (!this.metadata.category.includes("FUNCTION")) return;

        const child = this._children[0];
        if (child.metadata.category !== "PRODUCT") return;
        if (child.metadata.type !== "TYPE") return;
        return child;
    }

    public get productSelection(): GmodNode | undefined {
        if (this._children.length != 1) return;
        if (!this.metadata.category.includes("FUNCTION")) return;

        const child = this._children[0];
        if (!child.metadata.category.includes("PRODUCT")) return;
        if (child.metadata.type !== "SELECTION") return;

        return child;
    }

    public isChild(node: GmodNode | string): boolean {
        return typeof node === "string"
            ? this.isChildFromCode(node)
            : this.isChildFromCode(node.code);
    }

    private isChildFromCode(code: string) {
        for (let i = 0; i < this._children.length; i++) {
            if (this._children[i].code == code) return true;
        }

        return false;
    }

    public equals(other?: GmodNode): boolean {
        if (!other) return false;
        if (!!this._location !== !!other._location) return false;
        return (
            this.code === other.code &&
            (this._location?.equals(other._location) ?? true)
        );
    }

    public toString(): string {
        return this._location ? `${this.code}-${this._location}` : this.code;
    }

    public addChild(child: GmodNode) {
        this._children.push(child);
    }

    public addParent(parent: GmodNode) {
        this._parents.push(parent);
    }

    private getSortedString(node: GmodNode) {
        return node._location ? node.code + "-" + node._location : node.code;
    }

    public get isLeafNode(): boolean {
        return Gmod.isLeafNode(this.metadata);
    }
    public get isFunctionNode(): boolean {
        return Gmod.isFunctionNode(this.metadata);
    }

    public get isFunctionComposition(): boolean {
        return (
            (this.metadata.category === "ASSET FUNCTION" &&
                this.metadata.type === "COMPOSITION") ||
            (this.metadata.category === "PRODUCT FUNCTION" &&
                this.metadata.type === "COMPOSITION")
        );
    }

    private with(u: { (state: GmodNode): void }): GmodNode {
        const n = this.clone();
        u && u(n);
        return n;
    }

    public clone(): GmodNode {
        return new GmodNode(this);
    }
}

export function isIndividualizable(
    node: GmodNode,
    isTargetNode: boolean,
    isInSet: boolean = false
): boolean {
    if (node.metadata.type === "GROUP") return false;
    if (node.metadata.type === "SELECTION") return false;
    if (node.isProductType) return false;
    if (node.metadata.category === "ASSET" && node.metadata.type == "TYPE")
        return false;
    if (
        node.metadata.category === "ASSET FUNCTION" &&
        node.metadata.type === "COMPOSITION"
    )
        return (
            node.code[node.code.length - 1] === "i" || isTargetNode || isInSet
        );
    if (
        node.metadata.category === "PRODUCT FUNCTION" &&
        node.metadata.type === "COMPOSITION"
    )
        return (
            node.code[node.code.length - 1] === "i" || isTargetNode || isInSet
        );
    return true;
}

File: python/vista_sdk/GmodNode.py
from __future__ import annotations

from dataclasses import dataclass, field, replace
from typing import List, Optional, Tuple

from .GmodDto import GmodNodeDto
from .Locations import Location
from .ParsingErrors import ParsingErrors
from .VisVersions import VisVersion


class GmodNodeMetadata:
    def __init__(
        self,
        category,
        type,
        name,
        common_name,
        definition,
        common_definition,
        install_substructure,
        normal_assignment_names,
    ):
        self.category = category
        self.type = type
        self.name = name
        self.common_name = common_name
        self.definition = definition
        self.common_definition = common_definition
        self.install_substructure = install_substructure
        self.normal_assignment_names = normal_assignment_names

    @property
    def full_type(self) -> str:
        return f"{self.category} {self.type}"


@dataclass
class GmodNode:
    vis_version: VisVersion
    code: str
    metadata: GmodNodeMetadata
    location: Optional[Location] = None
    children: List[GmodNode] = field(default_factory=list)
    parents: List[GmodNode] = field(default_factory=list)

    @staticmethod
    def create_from_dto(vis_version: VisVersion, dto: GmodNodeDto) -> GmodNode:
        metadata = GmodNodeMetadata(
            dto.category,
            dto.type,
            dto.name,
            dto.common_name,
            dto.definition,
            dto.common_definition,
            dto.install_substructure,
            dto.normal_assignment_names
            if dto.normal_assignment_names is not None
            else {},
        )
        return GmodNode(vis_version=vis_version, code=dto.code, metadata=metadata)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, GmodNode):
            return NotImplemented
        if self.code != other.code:
            return False
        if (
            self.location is not None
            and other.location is not None
            and self.location.value != other.location.value
        ):
            return False
        if self.location is None and other.location is not None:
            return False
        return not (self.location is not None and other.location is None)

    def without_location(self) -> GmodNode:
        return GmodNode(
            vis_version=self.vis_version,
            code=self.code,
            metadata=self.metadata,
            children=self.children,
            parents=self.parents,
        )

    def with_location(self, location_str: Optional[str]) -> GmodNode:
        node = self.try_with_location(location_str)
        if node:
            return node
        raise ValueError(f"Invalid location: {location_str}")

    def try_with_location(self, location_str: Optional[str]) -> GmodNode:
        from .VIS import VIS

        if location_str is None:
            return self
        locations = VIS().get_locations(self.vis_version)
        try:
            new_location = locations.try_parse(location_str)[1]
            return GmodNode(
                vis_version=self.vis_version,
                code=self.code,
                metadata=self.metadata,
                location=new_location,
                children=self.children,
                parents=self.parents,
            )
        except ValueError:
            return self

    def try_with_location_errors(
        self, location_str: Optional[str], errors: List[ParsingErrors]
    ) -> Tuple[GmodNode, List[ParsingErrors]]:
        from .VIS import VIS

        if location_str is None:
            return self, errors
        locations = VIS().get_locations(self.vis_version)
        if locations.try_parse(location_str):
            new_location = locations.parse(location_str)
            return GmodNode(
                vis_version=self.vis_version,
                code=self.code,
                metadata=self.metadata,
                location=new_location,
                children=self.children,
                parents=self.parents,
            ), errors
        return self, errors

    def is_individualizable(
        self, is_target_node: bool = False, is_in_set: bool = False
    ) -> bool:
        if self.metadata.type in ["GROUP", "SELECTION"]:
            return False
        if self.is_product_type:
            return False
        if self.metadata.category == "ASSET" and self.metadata.type == "TYPE":
            return False
        if self.is_function_composition:
            return self.code[-1] == "i" or is_in_set or is_target_node
        return True

    @property
    def is_function_composition(self) -> bool:
        return (
            self.metadata.category in ["ASSET FUNCTION", "PRODUCT FUNCTION"]
            and self.metadata.type == "COMPOSITION"
        )

    @property
    def is_mappable(self) -> bool:
        if (
            self.product_type
            or self.product_selection
            or self.is_product_selection
            or self.is_asset
        ):
            return False
        last_char = self.code[-1]
        return last_char not in ["a", "s"]

    @property
    def is_product_selection(self) -> bool:
        return self.metadata.category == "PRODUCT" and self.metadata.type == "SELECTION"

    @property
    def is_product_type(self) -> bool:
        return self.metadata.category == "PRODUCT" and self.metadata.type == "TYPE"

    @property
    def is_asset(self) -> bool:
        return self.metadata.category == "ASSET" and self.metadata.type in [
            "TYPE",
            "SELECTION",
        ]

    @property
    def product_type(self) -> Optional[GmodNode]:
        if len(self.children) == 1:
            child = self.children[0]
            if (
                "FUNCTION" in self.metadata.category
                and child.metadata.category == "PRODUCT"
                and child.metadata.type == "TYPE"
            ):
                return child
        return None

    @property
    def product_selection(self) -> Optional[GmodNode]:
        if len(self.children) == 1:
            child = self.children[0]
            if (
                "FUNCTION" in self.metadata.category
                and "PRODUCT" in child.metadata.category
                and child.metadata.type == "SELECTION"
            ):
                return child
        return None

    def clone(self, **changes) -> GmodNode:
        return replace(self, **changes)

    def add_child(self, child: GmodNode) -> None:
        if child not in self.children:
            self.children.append(child)

    def add_parent(self, parent: GmodNode) -> None:
        if parent not in self.parents:
            self.parents.append(parent)

    def is_child(self, node_or_code) -> bool:
        code = node_or_code if isinstance(node_or_code, str) else node_or_code.code
        return any(child.code == code for child in self.children)

    def is_leaf_node(self) -> bool:
        full_type: str = self.metadata.full_type
        return full_type in ["ASSET FUNCTION LEAF", "PRODUCT FUNCTION LEAF"]

    def is_function_node(self) -> bool:
        return self.metadata.category != "PRODUCT" and self.metadata.type != "ASSET"

    def is_asset_function_node(self) -> bool:
        return (
            self.metadata.category == "ASSET FUNCTION"
            and self.metadata.type == "FUNCTION"
        )

    def is_root(self) -> bool:
        return self.code == "VE"

    def __str__(self):
        return (
            f"{self.code}-{self.location}" if self.location is not None else self.code
        )
